File: src/day1/ArrayList.ts
export default class ArrayList<T> {
    public length: number;
    private array: T[];
    private capacity: number;

    constructor(capacity: number) {
        this.capacity = capacity;
        this.length = 0;
        this.array = new Array(this.capacity);
    }

    prepend(item: T): void {
        this.length++;

        // Increase the capacity of the array if current capacity is reached.
        if (this.length > this.capacity) {
            this.capacity = this.capacity * 2;
            let newArray = new Array(this.capacity);
            newArray[0] = item;

            // Copy the data over to the new array. And prepend the new item.
            for (let i = 1; i < this.length; i++) {
                newArray[i] = this.array[i - 1];
            }

            this.array = newArray;

            return;
        }

        // Shift the array one space to the right.
        for (let i = this.length - 1; i >= 0; i--) {
            this.array[i + 1] = this.array[i];
        }

        this.array[0] = item;
    }

    insertAt(item: T, idx: number): void {
        // CHAT GPT IMPLEMENTATION //
        if (idx < 0 || idx > this.length) {
            throw new Error("Index out of bounds.");
        }

        if (this.length === this.capacity) {
            this.capacity = this.capacity * 2;
            const newArray = new Array(this.capacity);

            for (let i = 0; i < idx; i++) {
                newArray[i] = this.array[i];
            }

            newArray[idx] = item;

            for (let i = idx; i < this.length; i++) {
                newArray[i + 1] = this.array[i];
            }

            this.array = newArray;
        } else {
            for (let i = this.length; i > idx; i--) {
                this.array[i] = this.array[i - 1];
            }

            this.array[idx] = item;
        }

        this.length++;
    }

    append(item: T): void {
        this.length++;

        // Increase the capacity of the array if current capacity is reached.
        if (this.length > this.capacity) {
            this.capacity = this.capacity * 2;
            let newArray = new Array(this.capacity);

            // Copy the data over to the new array.
            for (let i = 0; i < this.length; i++) {
                newArray[i] = this.array[i];
            }

            this.array = newArray;
        }

        this.array[this.length - 1] = item;
    }

    remove(item: T): T | undefined {
        let removedIdx: number | undefined = undefined;

        for (let i = 0; i < this.length; i++) {
            if (this.array[i] === item) {
                removedIdx = i;
                break;
            }
        }

        if (removedIdx !== undefined) {
            return this.removeAt(removedIdx);
        }

        return;
    }

    get(idx: number): T | undefined {
        return this.array[idx];
    }

    removeAt(idx: number): T | undefined {
        if (idx < 0 || idx >= this.length) {
            throw new Error("Index out of bounds.");
        }

        const value = this.array[idx];

        // Shift the values in the array from the index to the left.
        for (let i = idx; i < this.length; i++) {
            this.array[i] = this.array[i + 1];
        }

        this.length--;

        return value;
    }
}

File: src/day1GoLang/array_list/array_list.go
package arrayList

import (
	"errors"
)

type ArrayList struct {
	length int
	list   []int
}

func (a *ArrayList) append(val int) {
	a.length++

	a.list = append(a.list, val)
}

func (a *ArrayList) get(idx int) (int, error) {
	if idx < 0 || idx >= a.length {
		return 0, errors.New("index out of bounds")
	}

	return a.list[idx], nil
}

func (a *ArrayList) prepend(val int) {
	a.length++

	if a.length == 1 {
		a.list = append(a.list, val)
		return
	}

	// Shift the array one space to the right.
	a.list = append([]int{val}, a.list...)
}

func (a *ArrayList) remove(val int) (int, error) {
	var removedIdx int

	for i := 0; i < a.length; i++ {
		if a.list[i] == val {
			removedIdx = i
			break
		}
	}

	if removedIdx > 0 || a.list[removedIdx] == val {
		return a.removeAt(removedIdx)
	}

	return val, errors.New("value not found")
}

func (a *ArrayList) removeAt(idx int) (int, error) {
	if idx < 0 || idx >= a.length {
		return 0, errors.New("index out of bounds")
	}

	a.length--
	val := a.list[idx]
	a.list = append(a.list[:idx], a.list[idx+1:]...)

	return val, nil
}
